crates/metrics/src/lib.rs

```diff
@@ -2,40 +2,131 @@
 
 use serde_derive::{Deserialize, Serialize};
 use std::{collections::HashMap, fs, io, path::Path, time::Duration};
+use sysinfo::{CpuExt, System, SystemExt};
 use thiserror::Error;
 
-/// Cycle-count metrics for a particular workload.
-///
-/// Stores the total cycle count and a breakdown of cycle count per named region.
-#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
-pub enum WorkloadMetrics {
-    /// Metrics produced when benchmarking in execution mode
-    Execution {
-        /// Name of the workload (e.g., "fft", "aes").
-        name: String,
+/// Represents a single benchmark run.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct BenchmarkRun {
+    /// Name of the benchmark.
+    pub name: String,
+    /// Block used gas
+    pub block_used_gas: u64,
+    /// Execution metrics for the benchmark run.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub execution: Option<ExecutionMetrics>,
+    /// Proving metrics for the benchmark run.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub proving: Option<ProvingMetrics>,
+}
+
+/// Hardware specs of the benchmark runner.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct HardwareInfo {
+    /// CPU model name.
+    pub cpu_model: String,
+    /// Total RAM in GiB.
+    pub total_ram_gib: u64,
+    /// Available GPUs.
+    pub gpus: Vec<GpuInfo>,
+}
+
+/// Information about a GPU.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct GpuInfo {
+    /// GPU model name.
+    pub model: String,
+}
+
+impl HardwareInfo {
+    /// Detects hardware information from the current system.
+    pub fn detect() -> Self {
+        let mut system = System::new_all();
+        system.refresh_all();
+
+        HardwareInfo {
+            cpu_model: system
+                .cpus()
+                .first()
+                .map(|cpu| cpu.brand().to_string())
+                .unwrap_or_else(|| "Unknown CPU".to_string()),
+            total_ram_gib: system.total_memory() / (1024 * 1024 * 1024),
+            gpus: detect_gpus(),
+        }
+    }
+
+    /// Serializes the hardware information to a JSON string in the provided path.
+    pub fn to_path<P: AsRef<Path>>(&self, path: P) -> Result<(), MetricsError> {
+        let path = path.as_ref();
+        ensure_parent_dirs(path)?;
+        let json = serde_json::to_string_pretty(self)?;
+        fs::write(path, json)?;
+        Ok(())
+    }
+}
+
+/// Detects available GPUs on the system.
+fn detect_gpus() -> Vec<GpuInfo> {
+    let mut gpus = Vec::new();
+
+    if let Ok(output) = std::process::Command::new("nvidia-smi")
```

Comment:
If `nvidia-smi` isn't found, "no gpus" would be detected. I think for now should be fine, didn't want to get into more hairy error parsing.

Also note we only try detecting NVIDIA gpus here (compared to AMD or other). I think that's a reasonable assumption to for now.